polars/polars-ops/src/series/ops/rle.rs

```diff
@@ -0,0 +1,35 @@
+use polars_core::prelude::*;
+
+pub fn rle(s: &Series) -> PolarsResult<Series> {
+    let (s1, s2) = (s.slice(0, s.len() - 1), s.slice(1, s.len()));
+    let s_neq = s1.not_equal_missing(&s2)?;
+    let n_runs = s_neq.sum().unwrap() + 1;
+    let mut lengths = Vec::with_capacity(n_runs as usize);
+    lengths.push(1);
+    let mut vals = Series::new_empty("values", s.dtype());
+    let vals = vals.extend(&s.head(Some(1)))?.extend(&s2.filter(&s_neq)?)?;
+    let mut idx = 0;
+    for v in s_neq.into_iter() {
+        if v.unwrap() {
+            idx += 1;
+            lengths.push(1);
+        } else {
+            lengths[idx] += 1;
+        }
+    }
+
+    let outvals = vec![Series::from_vec("lengths", lengths), vals.to_owned()];
+    Ok(StructChunked::new("rle", &outvals)?.into_series())
+}
+
+pub fn rle_id(s: &Series) -> PolarsResult<Series> {
+    let (s1, s2) = (s.slice(0, s.len() - 1), s.slice(1, s.len()));
+    // Run numbers start at zero
+    Ok(std::iter::once(false)
```

Comment:
We can speed this up by not using chains and flattens, but directly `extend` into a preallocated `Vec`. 

This will save a lot of branches and ensures we only allocate once.

Reply:
I thought the iterators would be efficient, but I guess you have to allocate somewhere. And your version is much simpler to read.

Reply:
The chain iterator is force to branch on both iterators. This branch can block other optimizations. And the iterators lenght becomes unknown, which leads to wrong allocations sizes and memcopys on realloc.

Reply:
Oh, gross. I didn't know that. I guess iterators aren't as "free" as they're made out to be.